crates/cli/src/main_shim.rs

```diff
@@ -0,0 +1,211 @@
+
+    if let Json::Array(after_args) = &shim["after_args"] {
+        for arg in after_args {
+            if let Json::Str(arg) = arg {
+                passthrough_args.push(OsString::from(arg));
+            }
+        }
+    }
+
+    // Create a command for local testing
+    // let mut command = Command::new("node");
+    // command.arg("./docs/shim-test.mjs");
+    // command.arg("--version");
+
+    // Create the command and handle alternate logic
+    let mut command = Command::new(get_proto_binary(&proto_home_dir, shim_exe_path));
+
+    if let Json::Str(parent_name) = &shim["parent"] {
+        command.args(["run", parent_name]);
+
+        if matches!(shim["alt_bin"], Json::Bool(true)) {
+            command.args(["--alt", shim_name]);
+        }
+    } else {
+        command.args(["run", shim_name]);
+    }
+
+    if !passthrough_args.is_empty() {
+        command.arg("--");
+        command.args(passthrough_args);
+    }
+
+    if let Json::Object(env_vars) = &shim["env_vars"] {
+        for (env, value) in env_vars {
+            if let Json::Str(var) = value {
+                command.env(env, var);
+            }
+        }
+    }
+
+    Ok(command)
+}
+
+pub fn main() -> Result<()> {
+    sigpipe::reset();
+
+    // Setup tracing and pass log level down
+    let log_level = env::var("PROTO_LOG").unwrap_or_else(|_| "info".into());
+
+    tracing::setup_tracing(TracingOptions {
+        filter_modules: vec!["proto".into()],
+        intercept_log: env::var("PROTO_WASM_LOG").is_err(),
+        log_env: "PROTO_LOG".into(),
+        ..TracingOptions::default()
+    });
+
+    // Extract arguments to pass-through
+    let args = env::args_os().collect::<Vec<_>>();
+    let exe_path = env::current_exe().unwrap_or_else(|_| PathBuf::from(&args[0]));
+
+    let shim_name = exe_path
+        .file_name()
+        .map(|file| String::from_utf8_lossy(file.as_encoded_bytes()))
+        .unwrap_or_default()
+        .replace(".exe", "");
+
+    trace!(shim = &shim_name, shim_bin = ?exe_path, args = ?args,  "Running {} shim", shim_name);
+
+    if shim_name.is_empty() || shim_name.contains("proto-shim") {
+        return Err(anyhow!(
+            "Invalid shim name detected. Unable to execute the appropriate proto tool.\nPlease refer to the documentation or ask for support on Discord."
+        ));
+    }
+
+    // Create the command to execute
+    let mut command = create_command(args, &shim_name, &exe_path)?;
+    command.env("PROTO_LOG", log_level);
+
+    // Spawn a shareable child process
+    trace!(
+        shim = &shim_name,
+        proto_bin = ?command.get_program(),
+        "Spawning proto child process"
+    );
+
+    let shared_child = SharedChild::spawn(&mut command)?;
+    let child = Arc::new(shared_child);
+    let child_clone = Arc::clone(&child);
+
+    // Handle CTRL+C and kill the child
+    ctrlc::set_handler(move || {
+        trace!("Received ctrl + c, killing child process");
+        let _ = child_clone.kill();
```

Comment:
I don't understand why this is necessary (at least on macOS and Linux). It has the adverse effect of altering the exit code of the subprocess from 130 to 1.

```
$ timeout --preserve-status --signal=INT 0.2 node; echo "exit code=$?"
...
exit code=1
$ timeout --preserve-status --signal=INT 0.2 $(proto bin node); echo "exit code=$?"
...
exit code=130
```

Reply:
On the other hand, if the shim process receives a signal, but not the child, we aren't forwarding it to the child process.

I'm doing this with `timeout --foreground`. Can you tell that I'm having fun with my new sledgehammer?

e.g. Launching node directly and signaling it causes a graceful exit
```ShellSession
$ timeout --signal=QUIT --foreground 1 $(proto bin node)
Welcome to Node.js v20.10.0.
Type ".help" for more information.
>
$
```

e.g. This kills the shim but not `node`, causing `node` to subsequently crash:
```ShellSession
$ timeout --signal=QUIT --foreground 1 node
Welcome to Node.js v20.10.0.
Type ".help" for more information.
>
$
node:events:492                                                                 
      throw er; // Unhandled 'error' event
      ^

Error: read EIO
    at TTY.onStreamRead (node:internal/stream_base_commons:217:20)
    at TTY.callbackTrampoline (node:internal/async_hooks:130:17)
Emitted 'error' event on REPLServer instance at:
    at ReadStream.onerror (node:internal/readline/interface:239:10)
    at ReadStream.emit (node:events:514:28)
    at ReadStream.emit (node:domain:488:12)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -5,
  code: 'EIO',
  syscall: 'read'
}

Node.js v20.10.0
```

Reply:
I've always been under the impression that ctrl+c'ing the parent process does not kill the child process, and it becomes a detached/zombie process. This has been a big problem in moon. https://doc.rust-lang.org/std/process/struct.Child.html

Let me test some more.

Reply:
I think to solve this correctly, we somehow need to wrap these in a process group, at least on unix.

Reply:
This is a good thread: https://github.com/rust-lang/cargo/issues/2343

Reply:
I am well out of my depth here, but I don’t know that a process group is the thing to do here.

I know that if we *do* spawn a subprocess, that we can’t just let the default signal handlers bring down the parent process (hence why the ctrl-c handler helps - it suppresses the default behavior, but we need to deal with other signals like `SIGQUIT`).

If we don’t actually need to do anything after `node` ends, we can dodge the whole question by using `execve` on macOS and Linux. Maybe even [Windows](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execve-wexecve) - I’m not sure what it actually does there, but I bet it’s close enough to work!

Reply:
I tried 3 implementations, and only the `ctrlc` crate actually killed the child processes.

- Cargos impl: https://github.com/rust-lang/cargo/blob/master/crates/cargo-util/src/process_builder.rs
- https://crates.io/crates/ctrlc
- https://crates.io/crates/command-group

Reply:
Looks like that cargo link was coming to the same conclusions I was moseying towards :-).

I particularly like `ProcessBuilder::exec_replace`, and I think it gives a simple effective implementation:

[exec_replace.patch](https://github.com/moonrepo/proto/files/13699301/exec_replace.patch)

> I tried 3 implementations, and only the ctrlc crate actually killed the child processes.

By design, ctrl-c sends an interrupt signal, which the process can choose to block, handle, or ignore. Depending on the test setup, you might have been checking the wrong behavior - the check should be that the child process was delivered the intended signal (and only if the signal caused the child process to end should the parent process stop supervising it).

Reply:
I've been testing these outside of actually tests and running them manually.

I think another part of the problem is that there area actually 2 levels deep of child processes. That may be what's causing some confusion. It basically goes `node shim (1) -> proto run node (2) -> bin/node (3)`, and only 1 and 2 actually are handled correctly. Let me dig a further and maybe abstract all of this onto the same implementation.

Reply:
Got a lot of this working here: https://github.com/moonrepo/proto/pull/346